lib/active_model/serializer/attributes.rb

```diff
@@ -1,55 +1,23 @@
 module ActiveModel
   class Serializer
     module Attributes
-      # @api private
-      class Attribute
-        delegate :call, to: :reader
-
-        attr_reader :name, :reader
-
-        def initialize(name)
-          @name = name
-          @reader = :no_reader
-        end
-
-        def self.build(name, block)
-          if block
-            AttributeBlock.new(name, block)
-          else
-            AttributeReader.new(name)
-          end
-        end
-      end
-      # @api private
-      class AttributeReader < Attribute
-        def initialize(name)
-          super(name)
-          @reader = ->(instance) { instance.read_attribute_for_serialization(name) }
-        end
-      end
-      # @api private
-      class AttributeBlock < Attribute
-        def initialize(name, block)
-          super(name)
-          @reader = ->(instance) { instance.instance_eval(&block) }
-        end
-      end
-
       extend ActiveSupport::Concern
 
       included do
         with_options instance_writer: false, instance_reader: false do |serializer|
-          serializer.class_attribute :_attribute_mappings # @api private : maps attribute key names to names to names of implementing methods, @see #attribute
-          self._attribute_mappings ||= {}
+          serializer.class_attribute :_attributes_data # @api private
+          self._attributes_data ||= {}
```

Comment:
I wrote above:

> - in both cases the `Attribute` can get the serialization value given a block or record's attribute name, and then
> - whether or not the `Attribute` should *also* know the serialization key used for its value/

So, the difference in the data structure is that given a serializer containing

```ruby
attribute :object_attribute
attribute :aliased_attribute, :attribute_alias
attribute :dynamic_attribute { 'much_dynamic' }
```

The difference in _mappings vs. _data key/value pairs would be

| `_attribute_mappings` | `_attributes_data` |
|------------------------------ |-------------------------|
| `object_attribute: { name: serializer.read_attribute_for_serialization(:object_attribute) }` | `object_attribute: { key: :object_attribute, name: serializer.read_attribute_for_serialization(:object_attribute) }`
| `attribute_alias: { name: :aliased_attribute, value: serializer.read_attribute_for_serialization(:aliased_attribute) }` | `aliased_attribute: { key: :attribute_alias, name: :aliased_attribute, value: serializer.read_attribute_for_serialization(:aliased_attribute) }`
| `dynamic_attribute: { name: :dynamic_attribute, value: serializer.instance_eval(&block) }` | `dynamic_attribute: { key: :dynamic_attribute, name: :dynamic_attribute, value: serializer.instance_eval(&block) }` |


both of which resolve to:

```ruby
{
  object_attribute: object.send(:object_attribute),
  attribute_alias: object.send(:aliased_attribute),
  dynamic_attribute: 'dynamic_attribute'
}
```

by applying _attributes to either the _mappings or _data where they are:

| ---- | for `_attribute_mappings` | for `_attributes_data` |
|------------------------------ |-------------------------|----------------------|
| `_attributes` | `_attribute_mappings.keys #=> [:object_attribute, :attribute_alias, :dynamic_attribute]` |  `_attributes_data.values.map(&:key) #=> [:object_attribute, :aliased_attribute, :dynamic_attribute]`

which means that the definition of `_attributes` has changed when an attribute specifies `:key`; it would no longer be the serialization key, but the object attribute name.

I think this is why in every case I see, the `_attributes_data` ignores the key in each pair and calls `_attributes_data.values`, which is also why I called it a mapping. (Which I think I got from a comment on the original PR, IIRC).

That makes sense? That's the heart of what I was really asking about.